### PR TITLE
Replace ChunkEvent.Generated hook (fixes #3582)

### LINKED ISSUE #3582: ChunkEvent.Generated fires for already generated chunks
**I am currently running**
- SpongeVanilla version: Latest 1.18.1 API 9 build (issue is presumably also present in 1.17.1 API 8 builds, however I've only tested 1.18.1)
- Java version: 17
- Operating System: Linux
- Plugins/Mods: test plugin / N/A

**Issue Description**
ChunkEvent.Generated seems to fire randomly and for already generated chunks.
A plugin that logs the chunk coordinates each time the event fires, and some tests like generating a world and loading an already generated world demonstrates this easily.

### PR DESCRIPTION
**Pull request by jpenilla**
fixes #3582

### CODE REVIEW

File: src/mixins/java/org/spongepowered/common/mixin/api/minecraft/server/level/ChunkHolderMixin_API.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.common.SpongeCommon;
+import org.spongepowered.common.event.ShouldFire;
+import org.spongepowered.common.event.tracking.PhaseTracker;
+import org.spongepowered.common.util.VecHelper;
+import org.spongepowered.math.vector.Vector3i;
+
+@Mixin(ChunkHolder.class)
+public abstract class ChunkHolderMixin_API {
```
Comment (gabizou) [inline src/mixins/java/org/spongepowered/common/mixin/api/minecraft/server/level/ChunkHolderMixin_API.java:44]: This isn't quite in the right package. This is implementing an event for the API, but not implementing any api interfaces itself.

File: src/mixins/java/org/spongepowered/common/mixin/api/minecraft/server/level/ChunkHolderMixin_API.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @Inject(
+            method = "replaceProtoChunk(Lnet/minecraft/world/level/chunk/ImposterProtoChunk;)V",
+            at = @At("TAIL")
+    )
+    private void impl$chunkGenerated(final ImposterProtoChunk imposter, final CallbackInfo ci) {
+        if (ShouldFire.CHUNK_EVENT_GENERATED) {
```
Comment (gabizou) [inline src/mixins/java/org/spongepowered/common/mixin/api/minecraft/server/level/ChunkHolderMixin_API.java:50]: I almost prefer a fail fast (it compiles down to the same thing anyways) like
```java
if (!ShouldFire.CHUNK_EVENT_GENERATED) {
  return;
}
// throw event
```

File: testplugins/src/main/java/org/spongepowered/test/chunk/ChunkEventTest.java [hunk 3]
```diff
@@ -60,28 +62,32 @@ public void enable(final CommandContext ctx) {
...
+            this.logChunkEvent("Post Save", event.chunkPosition(), event.worldKey());
         }
 
         @Listener
         public void onChunkUnload(final ChunkEvent.Unload event) {
-            Sponge.game().systemSubject().sendMessage(Component.text("Unload Chunk " + event.chunkPosition() + " in " + event.worldKey().asString()));
+            this.logChunkEvent("Unload", event.chunkPosition(), event.worldKey());
+        }
+
+        private void logChunkEvent(final String name, final Vector3i chunkPos, final ResourceKey worldKey) {
```
Comment (gabizou) [inline testplugins/src/main/java/org/spongepowered/test/chunk/ChunkEventTest.java:89]: Doesn't need to be an instance method, right?

**Review by gabizou — COMMENTED** [review]
Some nitpicks, but otherwise good stuff. I'll get to merging this if the notes aren't fixed (relocating the mixing to the appropriate folder/mixin set.

**Comment by jpenilla** [comment]
Looks like the fix is merged for 8 and 9 now, did you still want me to update this PR with the changes for the test plugin?

**Comment by gabizou** [comment]
Nah, the fix was merged, I forgot to close with the commit message. Merged/squashed in https://github.com/SpongePowered/Sponge/commit/7e9483279bd926ba62569136941c94fed739fc2e

I do appreciate the work put in and identifying the cause!

**Pull request closed without merging**